src/cascadia/TerminalSettingsModel/FileUtils.cpp

```diff
@@ -0,0 +1,127 @@
+            // the new AppModel paths (Packages/xxx/RoamingState, etc.) for standard path requests.
+            // Using this flag allows us to avoid Windows.Storage.ApplicationData completely.
+            THROW_IF_FAILED(SHGetKnownFolderPath(FOLDERID_LocalAppData, KF_FLAG_FORCE_APP_DATA_REDIRECTION, nullptr, &localAppDataFolder));
+
+            std::filesystem::path parentDirectoryForSettingsFile{ localAppDataFolder.get() };
+
+            if (!IsPackaged())
+            {
+                parentDirectoryForSettingsFile /= UnpackagedSettingsFolderName;
+            }
+
+            // Create the directory if it doesn't exist
+            std::filesystem::create_directories(parentDirectoryForSettingsFile);
+
+            return parentDirectoryForSettingsFile;
+        }();
+        return baseSettingsPath;
+    }
+
+    std::string ReadUTF8File(const std::filesystem::path& path)
+    {
+        // From some casual observations we can determine that:
+        // * ReadFile() always returns the requested amount of data (unless the file is smaller)
+        // * It's unlikely that the file was changed between GetFileSize() and ReadFile()
+        // -> Lets add a retry-loop just in case, to not fail if the file size changed.
+        for (int i = 0; i < 3; ++i)
+        {
+            wil::unique_hfile file{ CreateFileW(path.c_str(), GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, nullptr, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, nullptr) };
+            THROW_LAST_ERROR_IF(!file);
+
+            const auto fileSize = GetFileSize(file.get(), nullptr);
+            THROW_LAST_ERROR_IF(fileSize == INVALID_FILE_SIZE);
+
+            // By making our buffer just slightly larger we can detect if
+            // the file size changed and we've failed to read the full file.
+            std::string buffer(static_cast<size_t>(fileSize) + 1, '\0');
+            DWORD bytesRead = 0;
+            THROW_LAST_ERROR_IF(!ReadFile(file.get(), buffer.data(), gsl::narrow<DWORD>(buffer.size()), &bytesRead, nullptr));
+
+            // This implementation isn't atomic as it's not trivially possible to
+            // atomically read the entire file without breaking workflow of our users.
+            // For instance locking the file would conflict with users who have the file open in an editor.
+            // We'll just throw in case where the file size changed as that's likely an error.
+            if (bytesRead != fileSize)
+            {
+                // This continue is unlikely to be hit (see the prior for loop comment).
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                continue;
+            }
+
+            // As mentioned before our buffer was allocated oversized.
+            buffer.resize(bytesRead);
+
+            if (til::starts_with(buffer, Utf8Bom))
+            {
+                buffer.erase(0, Utf8Bom.size());
+            }
+
+            return buffer;
+        }

```

Comment:
The retry-loop is basically just there to retry in case that `ReadFile` failed to read the entire file.
It can happen that an application writes to the file we're currently reading. We can't actually detect this while simultaneously allowing users to have the file open in VS Code for instance, as this would require us to lock the file.
So what I did instead is to do the next best thing: If the file size changed between calling `GetFileSize` and `ReadFile`, we can be 100% sure that someone modified the file inbetween. The file could be corrupt, or we could've only read half the file. For such cases I built the retry-loop as a soft fallback. Currently we ignore such race conditions with other applications basically entirely.